server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java

```diff
@@ -0,0 +1,686 @@
+     *
+     * We then generally know very little, except that the safeVersion is a minimum version.
+     *
+     * Notice that since we only do CAS updates, we can sensibly handle this. If another CAS succeeds after this, we know that the
+     * previous CAS will not be able to succeed. Beware of the "next seqNo" handling in CASUpdateThread, which is necessary to be able to
+     * handle failed operations correctly here.
+     */
+    private static class FailState extends State {
+
+        private FailState(Version version) {
+            super(version);
+        }
+
+        @Override
+        Optional<State> casSuccess(Version inputVersion, Version outputVersion) {
+            if (safeVersion.gt(inputVersion) || safeVersion.gt(outputVersion))
+                return Optional.empty();
+
+            // the previous write could have been accepted or rejected so we cannot validate the version more precisely.
+            // but we know that any previous writes cannot succeed, since they all use seqNo <= inputVersion.seqNo.
+            return Optional.of(new SuccessState(outputVersion));
+        }
+
+        @Override
+        Optional<State> casFail(Version inputVersion, Version outputVersion) {
+            // we still do not know if the failed write could sneak in so have to stay in FailState.
+            return Optional.of(this);
+        }
+
+        @Override
+        Optional<State> fail() {
+            return Optional.of(this);
+        }
+    }
+
+
+    /**
+     * HistoryOutput serves both as the output of calls and delegating the sequential spec to the right State methods.
+     */
+    private interface HistoryOutput extends Serializable {
+        Optional<State> nextState(State currentState, Version input);
+
+        Version getVersion();
+    }
+
+    private static class IndexResponseHistoryOutput implements HistoryOutput {
+        private final Version outputVersion;
+
+        private IndexResponseHistoryOutput(IndexResponse response) {
+            this.outputVersion = new Version(response.getPrimaryTerm(), response.getSeqNo());
+        }
+
+        @Override
+        public Optional<State> nextState(State currentState, Version input) {
+            return currentState.casSuccess(input, outputVersion);
+        }
+
+        @Override
+        public Version getVersion() {
+            return outputVersion;
+        }
+
+        @Override
+        public String toString() {
+            return "Index{" + outputVersion + "}";
+        }
+    }
+
+    private static class CASFailureHistoryOutput implements HistoryOutput {
+        private Version outputVersion;
+        private CASFailureHistoryOutput(VersionConflictEngineException exception) {
+            this.outputVersion = parseException(exception.getMessage());
+        }
+
+        private Version parseException(String message) {
+            // ugly, but having these observed versions available improves the linearizability checking and ensures progress.
```

Comment:
Improved the comment in dc06ead10d4801037595901feeaba182b86a4d5a .

Using the version from the CAS failure exception ensures progress, since otherwise we would never be able to progress after just one CAS fail on own write.